cmds/df/df.go

```diff
@@ -0,0 +1,153 @@
+//
+// df reports details of mounted filesystems
+//
+// Synopsis
+//  df [-k] [-m]
+//
+// Description
+//  read mount information from /proc/mounts and
+//  statfs syscall and display summary information for all
+//  mount points that have a non-zero block count.
+//  Users can choose to see the diplay in KB or MB.
+//
+// Options
+//  -k: display values in KB (default)
+//  -m: dispaly values in MB
+//
+package main
+
+import (
+	"bytes"
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"math"
+	"syscall"
+)
+
+var (
+	inKB  = flag.Bool("k", false, "Express the values in kilobytes (default)")
+	inMB  = flag.Bool("m", false, "Express the values in megabytes")
+	units uint64
+)
+
+const mountinfoFile = "/proc/mounts"
+
+const (
+	// B is Bytes
+	B = 1
+	// KB is kilobytes
+	KB = 1024 * B
+	// MB is megabytes
+	MB = 1024 * KB
+)
+
+// Mount is a structure used to contain mount point data
+type Mount struct {
+	Device         string
+	MountPoint     string
+	FileSystemType string
+	Flags          string
+	Bsize          int64
+	Blocks         uint64
+	Total          uint64
+	Used           uint64
+	Avail          uint64
+	PCT            uint8
+}
+
+type mountinfomap map[string]Mount
+
+// mountinfo returns a map of mounts representing
+// the data in /proc/mounts
+func mountinfo() (mountinfomap, error) {
+	buf, err := ioutil.ReadFile(mountinfoFile)
+	if err != nil {
+		return nil, err
+	}
+	return mountinfoFromBytes(buf)
+}
+
+// returns a map generated from the bytestream returned
+// from /proc/mounts
+// for tidiness, we decide to ignore filesystems of size 0
+func mountinfoFromBytes(buf []byte) (mountinfomap, error) {
+	ret := make(mountinfomap, 0)
+	for _, line := range bytes.Split(buf, []byte{'\n'}) {
+		kv := bytes.SplitN(line, []byte{' '}, 6)
+		if len(kv) != 6 {
+			// can't interpret this
+			continue
+		}
+		key := string(kv[1])
+		var mnt Mount
+		mnt.Device = string(kv[0])
+		mnt.MountPoint = string(kv[1])
+		mnt.FileSystemType = string(kv[2])
+		mnt.Flags = string(kv[3])
+		DiskUsage(&mnt)
+		if mnt.Blocks == 0 {
+			continue
+		} else {
+			ret[key] = mnt
+		}
+	}
+	return ret, nil
+}
+
+// DiskUsage calculates the usage statistics of a mount point
+// Bsize on arm is int32, hence the cast
+func DiskUsage(mnt *Mount) {
+	fs := syscall.Statfs_t{}
+	err := syscall.Statfs(mnt.MountPoint, &fs)
```

Comment:
Hm, I wonder what `df` does in this case. It's possible for file systems to share a mountpoint -- e.g. I could mount tmpfs on /tmp, and then mount an ext4 volume over /tmp. The ext4 mount would be the only one visible., and a statfs call would show values for the ext4 mount.

This is kind of a strange case, though, so don't worry about solving it if you don't have any direct ideas.

Reply:
That's a great question!  

On investigating, I actually think linux df is a inconsistent here. [perhaps I mean the kernel is] 
I tried this:  
1. Open a shell and cd to /tmp.  run df
2. Open another shell and mount a disk filesystem over /tmp.  run df
3. Then go back to the first shell and run df.

In the first shell, if I write out a file, it goes into tmpfs, since it still has a handle on the lower filesystem.  The second shell can only see the higher (disk based) filesystem.

But the linux df doesn't follow that logic.
1. returns the size of tmpfs
2. returns the disk (of course)
3. ALSO returns the disk

So the first shell does not actually see the disk space available for it to write to (tmpfs size) - instead it's shown the disk size.
(At least that's true on my Arch system.  Other *nixes may vary)

 I'll investigate if I can actually do better or not, but I don't think I can.  I tried /proc/<pid>/mount{info} and see no difference in the responses...

Reply:
I think /proc/<pid> files are designed to work correctly with chroots rather than filesystems mounted over the top of each other.  
I think my example would be considered an 'edge case'; df is correct for 'normal' conditions - anyone beginning to write to that filesystem would see the correct size.  
Only processes left behind after the mount would continue to write to the old filesystem, which could be considered an administrative error - although I admit that's a opinionated view...
just my 2 pence, so happy to discuss.

Reply:
No worries, if upstream df doesn't do this right, we don't have to either. For now, this works and is better than the status quo.